### PR TITLE
Kylin 5256 Add a cache for the system property get by the optional config in KylinConfigBase

### PR DESCRIPTION
**Pull request by zznlime**
## Proposed changes

Add a cache for the system property get by the optional config in KylinConfigBase

## Branch to commit
- [ ] Branch **kylin3** for v2.x to v3.x
- [ ] Branch **kylin4** for v4.x
- [x] Branch **kylin5** for v5.x

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-5256 Make Kylin project open-source"
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by zznlime** [comment]
@hit-lacus

File: src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 1]
```diff
@@ -111,6 +110,33 @@ public abstract class KylinConfigBase implements Serializable {
 
     protected static final Map<String, String> STATIC_SYSTEM_ENV = new ConcurrentHashMap<>(System.getenv());
 
+    // It's a workaround to avoid lock in bottom hash table
+    // It can be removed after updating JDK to 11
+    protected static final ConcurrentHashMap<Object, Object> STATIC_SYSTEM_PROPERTY = new ConcurrentHashMap<>(
+            System.getProperties());
+
+    protected static String getSystemProperty(String key) {
+        Object oval = STATIC_SYSTEM_PROPERTY.get(key);
```
Comment (Mukvin) [inline src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:119]: system.getProperty() used the checkKey() for fast failed, so this place may check too.

File: src/core-common/src/main/java/org/apache/kylin/common/KylinExternalConfigLoader.java [hunk 2]
```diff
@@ -161,11 +163,18 @@ public String getConfig() {
 
     @Override
     public String getProperty(String key) {
-        return properties.getProperty(key);
+        Object oval = propertyEntries.get(key);
+        return (oval instanceof String) ? (String) oval : null;
```
Comment (Mukvin) [inline src/core-common/src/main/java/org/apache/kylin/common/KylinExternalConfigLoader.java:167]: recommend to use String.valueOf(oval)

File: src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java [hunk 3]
```diff
@@ -113,7 +114,7 @@ public static void restoreAllSystemProp(Map<String, String> systemProp) {
 
     /** Set system property */
     public static String setProperty(String property, String value) {
-        return System.setProperty(property, value);
+        return KylinConfigBase.setSystemProperty(property, value);
```
Comment (Mukvin) [inline src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java:117]: +1

File: src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java [hunk 4]
```diff
@@ -29,6 +29,7 @@
 import org.apache.commons.lang.StringUtils;
 
 import lombok.extern.slf4j.Slf4j;
+import org.apache.kylin.common.KylinConfigBase;
```
Comment (Mukvin) [inline src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java:32]: this import is weird, KylinConfigBase is an instance class, and KylinConfigBase can use Unsafe util, but Unsafe should not depend on KylinConfigBase.

File: src/core-common/src/main/java/org/apache/kylin/common/PropertiesDelegate.java [hunk 5]
```diff
@@ -18,36 +18,38 @@
...
+        if (configLoader != null)
+            this.properties.putAll(configLoader.getPropertyEntries());
         this.properties.putAll(properties);
         this.configLoader = configLoader;
     }
 
     public void reloadProperties(Properties properties) {
         this.properties.clear();
-        if(configLoader != null) this.properties.putAll(configLoader.getProperties());
+        if (configLoader != null)
```
Comment (hit-lacus) [inline src/core-common/src/main/java/org/apache/kylin/common/PropertiesDelegate.java:51]: If configLoader is instance of CachedNacosExternalConfigLoader, I think `configLoader.getPropertyEntries()` will return out of date values.

File: src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java (hunk 1 above)
Comment (zznlime) [reply src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:120]: Have added checkKey() for getSystemProperty.

File: src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 6]
```diff
@@ -138,15 +164,15 @@ public static String getKylinHome() {
...
+            kylinHome = getSystemProperty("KYLIN_HOME");
         }
         return kylinHome;
     }
 
     public static String getKylinConfHome() {
         String confHome = System.getenv("KYLIN_CONF");
         if (StringUtils.isEmpty(confHome)) {
-            confHome = System.getProperty("KYLIN_CONF");
+            confHome = getSystemProperty("KYLIN_HOME");
```
[the remark this answers is not in the document]
Comment (zznlime) [reply src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:176]: Fixed. Thanks for checking.

File: src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java (hunk 4 above)
Comment (zznlime) [reply src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java:32]: Have removed KylinConfigBase import from Unsafe.java

File: src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java (hunk 3 above)
Comment (zznlime) [reply src/core-common/src/main/java/org/apache/kylin/common/util/Unsafe.java:117]: Fixed

File: src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 7]
```diff
@@ -111,6 +110,33 @@ public abstract class KylinConfigBase implements Serializable {
...
+
+    protected static String getSystemProperty(String key, String defaultValue) {
+        String val = getSystemProperty(key);
+        return (val == null) ? defaultValue : val;
+    }
+
+    // Mainly invoked in tests
+    public static String setSystemProperty(String key, String value) {
+        System.setProperty(key, value);
+        return (String) STATIC_SYSTEM_PROPERTY.put(key, value);
```
[the remark this answers is not in the document]
Comment (zznlime) [reply src/core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:133]: this commit has been revert

**Pull request merged**